### DISCUSSION TITLE
Multiple connections to Nimble Server - how to automatically allow new connections to pair

### DISCUSSION THREAD

**Discussion by pwells13** [h2zero/NimBLE-Arduino#594] [Q&A]
Hi,
I have an application that allows up to 3 connections to a Nimble server. I am using a confirmation pin for pairing. What I would like to be able to do is to allow new connections to be made if less than 3 connections are currently in use. However, what I find is that once 3 connections have been made it is necessary to delete one of the bonds on the server before another connection can successfully pair. I'm struggling to see a way of doing this nicely, and in a reasonably automatic way. What works is to use the NimBLEDevice::deleteBond(NimBLEAddress) and get the NimBLEAddress from the NimBLEDevice::getBondedAddress(0) command, but then that bond could be used by a connection that is currently in use. Is there a way of matching the bond information with currently connected clients?

The reason for needing to control the bonded device list is that devices which have successfully paired and have a bond can then reconnect without having to go through the pairing process again.

For testing I have a simple sketch (attached) which I'm uploading to an ESP32 dev module, and then I'm using LightBlue to connect and listen to the characteristic. I have set up the ESP module so that a switch between pin 23 and ground acts as the confirmation for pairing (press this whilst the blue light on the dev board is on before clicking on pair on LightBlue). I have LightBlue running on a mix of iOs and Android devices.

When running the sketch, to delete bond(0) using the serial monitor in the Arduino IDE type any text in the input bar and press enter.

Hope this all makes sense! I am aware that I have lots to learn so any help to point me in the right direction would be much appreciated.

[NimBLE_ServerTesting2.txt](https://github.com/h2zero/NimBLE-Arduino/files/12857387/NimBLE_ServerTesting2.txt)

**Comment by h2zero** [comment]
This should work automatically as the nimble core library will delete the oldest paired device that is not connected. There may be a bug upstream if that is not the case.

**Comment by pwells13** [comment]
@h2zero 
Thank you for the reply. I was hoping that it would work automatically, but it seems that it doesn't in my case. I can connect additional devices once I have reached the limit (providing that the current number of devices actually connected at the time is below the limit), but I have to pair every time they connect as it seems the bond is not stored. If I delete one of the bonds then the new device can connect, pair and then reconnect without pairing. Is there any information I can pass on that would help in finding a solution, or can you point me in a direction where I might find a solution?

**Reply by h2zero** [reply]
This is all handled by the core nimble code. I'll have a look into this.

**Reply by pwells13** [reply]
Thank you! In the meantime I've got a work around using a function that I call in the onPinConfirm callback. It was a case of matching the stored bond addresses with any current connections to the server and then finding the first bond that wasn't in use.

**Reply by pwells13** [reply]
@h2zero 
Hi, just wondering if you had any further thoughts on this issue?
Many thanks

**Reply by h2zero** [reply]
I haven't had a chance to dig into this sorry. I've been very busy with work and I will be looking at this soon.
